src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -338,30 +392,70 @@ QueryPlanPtr MergeTreeDataSelectExecutor::read(
                     settings.compile_aggregate_expressions,
                     settings.min_count_to_compile_aggregate_expression);
 
-                transform_params = std::make_shared<AggregatingTransformParams>(
-                    std::move(params), aggregator_list_ptr, query_info.projection->aggregate_final);
-            }
-
-            pipe.resize(pipe.numOutputPorts(), true, true);
-
-            auto merge_threads = num_streams;
-            auto temporary_data_merge_threads = settings.aggregation_memory_efficient_merge_threads
-                ? static_cast<size_t>(settings.aggregation_memory_efficient_merge_threads)
-                : static_cast<size_t>(settings.max_threads);
+                bool only_merge = false;
+                if (projection)
+                {
+                    /// The source header is also an intermediate header
+                    params.intermediate_header = header_before_aggregation;
+
+                    /// This part is hacky.
+                    /// We want AggregatingTransform to work with aggregate states instead of normal columns.
+                    /// It is almost the same, just instead of adding new data to aggregation state we merge it with existing.
+                    ///
+                    /// It is needed because data in projection:
+                    /// * is not merged completely (we may have states with the same key in different parts)
+                    /// * is not split into buckets (so if we just use MergingAggregated, it will use single thread)
+                    only_merge = true;
+                }
 
-            pipe.addSimpleTransform([&](const Block & header)
-            {
-                return std::make_shared<AggregatingTransform>(
-                    header, transform_params, many_data, counter++, merge_threads, temporary_data_merge_threads);
-            });
-        };
+                auto merge_threads = num_streams;
+                auto temporary_data_merge_threads = settings.aggregation_memory_efficient_merge_threads
+                    ? static_cast<size_t>(settings.aggregation_memory_efficient_merge_threads)
+                    : static_cast<size_t>(settings.max_threads);
 
-        if (!projection_pipe.empty())
-            build_aggregate_pipe(projection_pipe, true);
-        if (!ordinary_pipe.empty())
-            build_aggregate_pipe(ordinary_pipe, false);
+                InputOrderInfoPtr group_by_info = query_info.projection->input_order_info;
```

Comment:
The code duplication is significant. If the `in_order_aggregate` optimization can only be applicated to one pipe, and if here `group_by_info` is the enabler, let's combine the two branches and only condition on this.

Reply:
In one case we need to add transform, in another step, that's why there are two different branches.
I will reduce copy paste a little though, and I'm open to any suggestions.